File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
from collections import defaultdict

def solution(operations):
    max_heap = []
    min_heap = []
    count_dict = defaultdict(int)
    
    for operation in operations:
        if operation.startswith("I"):
            # 삽입
            oper = operation.split(" ")
            num = int(oper[1])
            heapq.heappush(max_heap, -num)
            heapq.heappush(min_heap, num)
            count_dict[num] += 1
        elif operation.startswith("D -"):
            while min_heap:
                num = heapq.heappop(min_heap)
                if count_dict[num] > 0:
                    count_dict[num] -= 1
                    break
        else:
            while max_heap:
                num = -heapq.heappop(max_heap)
                if count_dict[num] > 0:
                    count_dict[num] -= 1
                    break
    
    if not min_heap or not max_heap:
        return [0,0]
    
    max_value = float('-inf')
    min_value = float('inf')
    
    while min_heap:
        num = heapq.heappop(min_heap)
        if count_dict[num] > 0:
            min_value = num
            break
    
    
    while max_heap:
        num = -heapq.heappop(max_heap)
        if count_dict[num] > 0:
            max_value = num
            break
    
    answer = [max_value, min_value]
    return answer

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;
import java.util.Map;
import java.util.HashMap;
import java.util.Comparator;


class Solution {
    public int[] solution(String[] operations) {
        Map<Integer, Boolean> validMap = new HashMap<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        
        for(String operation : operations){
            if(operation.startsWith("I")){
                String[] oper = operation.split(" ");
                int num = Integer.parseInt(oper[1]);
                validMap.put(num, true);
                maxHeap.add(num);
                minHeap.add(num);
            }else if(operation.equals("D -1")){
                // 최솟값 제거 
                while(!minHeap.isEmpty()){
                    int num = minHeap.poll();
                    if(validMap.get(num)){
                        validMap.put(num, false);
                        break;
                    }
                }
            }else if(operation.equals("D 1")){
                // 최댓값 제거
                while(!maxHeap.isEmpty()){
                    int num = maxHeap.poll();
                    if(validMap.get(num)){
                        validMap.put(num, false);
                        break;
                    }
                }
            }
        }
        
        int maxValue = 0;
        int minValue = 0;
        
        while(!minHeap.isEmpty()){
            int num = minHeap.poll();
            if(validMap.get(num)){
                minValue = num;
                break;
            }
        }
        
        while(!maxHeap.isEmpty()){
            int num = maxHeap.poll();
            if(validMap.get(num)){
                maxValue = num;
                break;
            }
        }
        
        
        
        int[] answer = {maxValue, minValue};
        return answer;
    }
}